Feature - Added Optional Google Analytics tag for Status Page.

clcninja opened the pull request:
- [x] I have read and understand the pull request rules.

# Description
This PR adds in a Google Analytics tag field for a Status Page (allowing different tags for different pages/domains).

Fixes #2180 (issue)

## Type of change
- User interface (UI)
- New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)
- [x] My changes generate no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)
<img width="328" alt="image" src="https://user-images.githubusercontent.com/54022576/211210254-a1ea6b1c-62f9-41d6-8347-b5c2fac0211b.png">

louislam commented on src/pages/StatusPage.vue:
```diff
@@ -294,6 +300,15 @@
+            window.dataLayer = window.dataLayer || [];
+            function gtag(){dataLayer.push(arguments);}
+            gtag('js', new Date());
+            gtag('config', '{{ config.googleAnalyticsId }}');
```
Need a better way to insert the code here. It will be syntax error if put something strange.

louislam commented:
I am thinking, maybe we should use sever side rendering instead. 

As the code have to put at beginning of `<head>`:
https://support.google.com/analytics/answer/1008080?hl=en#zippy=%2Cin-this-article

clcninja commented:
Sounds good, I'll refactor it into server/status_page.js

Saibamen commented on server/model/status_page.js:
```diff
@@ -225,6 +225,7 @@ class StatusPage extends BeanModel {
             customCSS: this.custom_css,
             footerText: this.footer_text,
             showPoweredBy: !!this.show_powered_by,
+            googleAnalyticsId: this.google_analytics_tag_id
```
```suggestion
            googleAnalyticsId: this.google_analytics_tag_id,
```

Saibamen commented on server/database.js:
```diff
@@ -66,6 +66,7 @@ class Database {
         "patch-add-radius-monitor.sql": true,
         "patch-monitor-add-resend-interval.sql": true,
         "patch-maintenance-table2.sql": true,
+        "patch-add-google-analytics-status-page-tag.sql": true
```
```suggestion
        "patch-add-google-analytics-status-page-tag.sql": true,
```

Saibamen commented on server/model/status_page.js:
```diff
@@ -245,6 +246,7 @@ class StatusPage extends BeanModel {
             customCSS: this.custom_css,
             footerText: this.footer_text,
             showPoweredBy: !!this.show_powered_by,
+            googleAnalyticsId: this.google_analytics_tag_id
```
```suggestion
            googleAnalyticsId: this.google_analytics_tag_id,
```

Saibamen reviewed (changes requested):
Missing new i18n keys

Computroniks commented on server/modules/google-analytics.js:
```diff
@@ -0,0 +1,11 @@
+let GoogleAnalytics = (() => {
+    function getGoogleAnalyticsScript(tagId) {
+        return "<script async src=\"https://www.googletagmanager.com/gtag/js?id=" + tagId + "\"></script>" +
+            "<script>window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date());gtag('config', '" + tagId + "'); </script>";
```
How is tagId sanitized when it is input. Is there anything stopping a user from entering something like 
```
<valid google tag>'); alert('Gotcha
```
or other, more questionable input?

clcninja replied:
I have now added a regex check to validate tags

louislam replied:
This is a XXS issue. It must be escaped on the rendering side, because if people reuse this function in the future without knowing it is vulnerable, it can be very dangerous.

I think [jsece](https://github.com/mathiasbynens/jsesc) should help to escape.

clcninja replied:
I've now added jsesc when rendering the script.

Computroniks commented on server/modules/google-analytics.js:
```diff
@@ -0,0 +1,16 @@
+let GoogleAnalytics = (() => {
+    function getGoogleAnalyticsScript(tagId) {
```
Would it be possible to add JSDoc to the undocumented methods

Computroniks commented on server/modules/google-analytics.js:
```diff
@@ -0,0 +1,16 @@
+        return "<script async src=\"https://www.googletagmanager.com/gtag/js?id=" + tagId + "\"></script>" +
+            "<script>window.dataLayer = window.dataLayer || []; function gtag(){dataLayer.push(arguments);} gtag('js', new Date());gtag('config', '" + tagId + "'); </script>";
+    }
+    function isValidTag(tagInput) {
```
And JSDoc here

clcninja replied:
Have added those as requested :)

louislam commented on server/modules/google-analytics.js:
```diff
@@ -0,0 +1,11 @@
+let GoogleAnalytics = (() => {
```
Also I think this module's instantiation is too fancy.

```javascript
function getGoogleAnalyticsScript(tagId) {
   return ...
}

modules.exports = {
   getGoogleAnalyticsScript,
}
```

I believe the example above do the same job.

clcninja replied:
Thanks for the example, I was looking at other modules and trying to fit in with other existing code.

louislam commented on server/model/status_page.js:
```diff
@@ -53,6 +54,15 @@ class StatusPage extends BeanModel {
 
         const head = $("head");
 
+        await StatusPage.getStatusPageData(statusPage).then( (page) => {
```
I think this call is not necessary, because the `statusPage` is representing the data row from database. I think you can get the UA from `statusPage.google_analytics_tag_id` directly.

clcninja replied:
Looks like you are correct! Updating the PR now! :)

louislam commented on db/patch-add-google-analytics-status-page-tag.sql:
```diff
@@ -0,0 +1,4 @@
+-- You should not modify if this have pushed to Github, unless it does serious wrong with the db.
+BEGIN TRANSACTION;
+ALTER TABLE status_page ADD google_analytics_tag_id TEXT;
```
Should be VARCHAR

clcninja replied:
Fixed ✔️

louislam commented:
My codes (GA3 and GA4) do not seem to be working, it is still empty after save.
- UA-36889556-1
- G-8Y5FKTE8S4

clcninja commented:
> My codes (GA3 and GA4) do not seem to be working, it is still empty after save.
> 
> * UA-36889556-1
> * G-8Y5FKTE8S4

Ah that's getting blocked by the regex, apparently that looks different than what the google ads site specifies, I'll make it a bit less restrictive, obviously they have used different formats at different points.

louislam commented:
> > My codes (GA3 and GA4) do not seem to be working, it is still empty after save.
> > 
> > * UA-36889556-1
> > * G-8Y5FKTE8S4
> 
> Ah that's getting blocked by the regex, apparently that looks different than what the google ads site specifies, I'll make it a bit less restrictive, obviously they have used different formats at different points.

Maybe remove this regex? Because Google will keep adding new code format which hard to handle in the future.

clcninja commented:
> > > My codes (GA3 and GA4) do not seem to be working, it is still empty after save.
> > > 
> > > * UA-36889556-1
> > > * G-8Y5FKTE8S4
> > 
> > 
> > Ah that's getting blocked by the regex, apparently that looks different than what the google ads site specifies, I'll make it a bit less restrictive, obviously they have used different formats at different points.
> 
> Maybe remove this regex? Because Google will keep adding new code format which will be hard to handle in the future.

A good shout I think. I've removed it and updated the database field type.

Saibamen commented on server/socket-handlers/status-page-socket-handler.js:
```diff
@@ -7,6 +7,7 @@ const Database = require("../database");
 const apicache = require("../modules/apicache");
 const StatusPage = require("../model/status_page");
 const { UptimeKumaServer } = require("../uptime-kuma-server");
+const googleAnalytics = require("../modules/google-analytics");
```
Not used

clcninja replied:
Fixed & pushed 👍

louislam commented on server/model/status_page.js:
```diff
@@ -53,6 +54,13 @@ class StatusPage extends BeanModel {
         const head = $("head");
 
+        if (statusPage.googleAnalyticsTagId) {
+            let escapedGoogleAnalyticsScript = jsesc(googleAnalytics.getGoogleAnalyticsScript(statusPage.googleAnalyticsTagId), {
```
I think jsesc is wrongly used, as `getGoogleAnalyticsScript()` returns with <script> tag which is not purely a js script. I will fix it
